### ISSUE TITLE
False negative for `dict` assignment with `Literal` keys

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#4276]
**Describe the bug**

When defining a module-level variable and typing it as `dict[Literal[...], T]`, the key type is respected in the assignment but will not report an error if any keys are missing from the definition, e.g.

```py
from typing_extensions import Literal

MyKey = Literal['foo', 'bar']

# this should report an error as the `foo` and `bar` keys are missing from the assignment
MAPPING: dict[MyKey, str] = {}
reveal_type(MAPPING)  # reveals correct type, `dict[MyKey, str]

# results in no error for the following line even though it will cause a runtime error
print(MAPPING['foo'])
```

**To Reproduce**

Run pyright in strict mode over the above script.

**Expected behavior**

An error should be reported for the `MAPPING` definition

**VS Code extension or command-line**

Command-line `1.1.282`

**Additional context**

I would expect this to have the same behaviour as if it were a generic `TypedDict`, e.g.

```py
class MyMapping(TypedDict, Generic[_T]):
    foo: _T
    bar: _T

# this correctly generates an error
MAPPING: MyMapping[str] = {}
```

**Comment by erictraut** [comment]
The behavior is correct. The The type `dict[MyKey, str]` means that it is a `dict` whose entries must have keys that are compatible with the type `Literal['foo', 'bar']`. It does not mean that such keys exist — just as `dict[str, str]` doesn't mean that a key exists for every possible `str` value.

If you want to indicate that entries with specific keys are required in your dictionary, you should use a `TypedDict` instead. That's what they we designed for.

```python
class MyMapping(TypedDict):
    foo: Required[str]
    bar: Required[str]

MAPPING: MyMapping = {} # Error: 'foo' and 'bar' are missing
```
